libbpf-tools/javagc.c

```diff
@@ -0,0 +1,237 @@
+#include <bpf/libbpf.h>
+#include <errno.h>
+#include "javagc.skel.h"
+#include "javagc.h"
+
+#define BINARY_PATH_SIZE (256)
+#define PERF_BUFFER_PAGES (32)
+#define PERF_POLL_TIMEOUT_MS (200)
+
+static struct env {
+	pid_t pid;
+	int time;
+	bool exiting;
+	bool verbose;
+} env = {
+	.pid = -1,
+	.time = 1000,
+	.exiting = false,
+	.verbose = false,
+};
+
+const char *argp_program_version = "javagc 0.1";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+
+const char argp_program_doc[] =
+"Monitor javagc time cost.\n"
+"\n"
+"USAGE: javagc [--help] [-p PID] [-t GC time]\n"
+"\n"
+"EXAMPLES:\n"
+"javagc -p 185         # trace PID 185 only\n"
+"javagc -p 185 -t 100  # trace PID 185 java gc time beyond 100us\n";
+
+static const struct argp_option opts[] = {
+	{ "pid", 'p', "PID", 0, "Trace this PID only" },
+	{ "time", 't', "TIME", 0, "Java gc time" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	int err = 0;
+
+	switch (key) {
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'p':
+		errno = 0;
+		env.pid = strtol(arg, NULL, 10);
+		if (errno) {
+			err = errno;
+			fprintf(stderr, "invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 't':
+		errno = 0;
+		env.time = strtol(arg, NULL, 10);
+		if (errno) {
+			err = errno;
+			fprintf(stderr, "invalid time: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return err;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && ! env.verbose)
+		return 0;
+
+	return vfprintf(stderr, format, args);
+}
+
+static void bump_memlock_rlimit(void)
+{
+	struct rlimit rlim_new = {
+		.rlim_cur	= RLIM_INFINITY,
+		.rlim_max	= RLIM_INFINITY,
+	};
+
+	if (setrlimit(RLIMIT_MEMLOCK, &rlim_new)) {
+		fprintf(stderr, "Failed to increase RLIMIT_MEMLOCK limit!\n");
+		exit(1);
+	}
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	struct data_t *e = (struct data_t *)data;
+	struct tm *tm = NULL;
+	char ts[16];
+	time_t t;
+	if (e->ts > env.time) {
+		time(&t);
+		tm = localtime(&t);
+		strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+		printf("time:%s, cpu:%d, pid:%d, gc time(us):%llu\n", ts, e->cpu,
+				e->pid, e->ts/1000);
+	}
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 data_sz)
+{
+	printf("lost data\n");
+}
+
+static void sig_handler(int sig)
+{
+	env.exiting = true;
+}
+
+static int get_binary_path(char *path)
+{
+	char buf[64] = {0};
+	FILE *f = NULL;
+
+	snprintf(buf, 64, "cat /proc/%d/maps | grep libjvm.so | awk '{print $6}' | uniq", env.pid);
+	f = popen(buf, "r");
+	if (!f)
+		return -1;
+	fscanf(f, "%s", path);
+	pclose(f);
+
+	return 0;
+}
```

Comment:
We've just landed a USDT example at libbpf-bootstrap repo.

https://github.com/libbpf/libbpf-bootstrap/blob/4f0d020eb92572bdf2a7606db4efa09e7fa37ea6/examples/c/usdt.c#L52-L53

libbpf should be able to do auto-attach, you don't need this.

Reply:
You mean i need not to find the path of the so binary, just set the so binary name?

Reply:
I see the source code of finding the binary. The libjvm.so seems not int the target path.

"""
static int resolve_full_path(const char *file, char *result, size_t result_sz)
{
	const char *search_paths[3] = {};
	int i;

	if (str_has_sfx(file, ".so") || strstr(file, ".so.")) {
		search_paths[0] = getenv("LD_LIBRARY_PATH");
		search_paths[1] = "/usr/lib64:/usr/lib";
		search_paths[2] = arch_specific_lib_paths();
	} else {
		search_paths[0] = getenv("PATH");
		search_paths[1] = "/usr/bin:/usr/sbin";
	}
"""